Add support for `Gradle` libraries including version incrementing, monorepos, and multi-module

adi-gnm opened the pull request:
<!-- See https://github.com/guardian/recommendations/blob/main/pull-requests.md for recommendations on raising and reviewing pull requests. -->

## What

This PR creates a reusable workflow for releasing Gradle based libraries. It starts with the existing `reusable_release.yml` from the Scala library release workflow, and adapts it to support:

1. Calculating versions for Gradle based libraries
2. Operating in monorepos where the Gradle libraries may be in a subfolder
3. Supporting multiple Gradle modules producing separate artifacts

I've also updated the CODEOWNERS file to include Android admins and Roberto for now.

I'll update the documentation, including integration steps in a follow up PR. My intent is to hand over the documentation and workflow to Daniel and ask him to integrate it in Coverdrop by himself. That would help me identify where the documentation is lacking.

## Why

This workflow improves the existing release implementation used by [Source](https://github.com/guardian/source-apps) & [Coverdrop](https://github.com/guardian/coverdrop):

1. Removes need for manually updating version numbers by automating it based on code changes
2. Adds one-click PREVIEW release support for testing WIP
3. Removes code duplication for versioning, signing, and publishing
4. Improves security of our Sonatype release secrets by isolating jobs that use them from jobs that run Gradle and use library code

## How

High level summary of changes:

1. Created `calculate-version.sh` to calculate the new version for the library
   This script reads the version in `version.txt` file and breaks it down into its MAJOR, MINOR and PATCH components. It then iterates over all input modules to determine the largest bump to version. Once the largest bump is determined, the previous version is bumped accordingly, written to `version.txt`, committed and tagged.
   
   For determining the version bump we use the existing `api.txt` file and the `metalavaCheckCompatibilityRelease` and `metalavaGenerateSignatureRelease` tasks:
   * If `metalavaCheckCompatibilityRelease` fails, we bump the MAJOR version, and reset others to zero
   * Else if `api.txt` changes after `metalavaGenerateSignatureRelease`, then we bump the MINOR version, and reset the patch to zero
   * Else we bump the PATCH version
2. Extracted and updated `create-unsigned-commit.sh` to create unsigned commits for the repo version file, and the API files for each module. It also sets all the API version file changes (path, sha, contents as base64) as job output by wrapping them in a json object.
   This is mostly identical to the existing job in the parent Scala workflow. Key changes being
   * extracted to a script file, 
   * added support for mutliple modules, and 
   * added support for committing `api.txt` files
3. Added new workflow inputs to support _mono-repo_ and _multi-module_ repos
   * `SOURCE_DIR` - optional path to the Gradle project root directory. Only needed for monorepos where the Gradle project may be in a subfolder.
   * `MODULES` - a comma separated list of module names for individual artifacts 
      * **Note:** Nested modules, e.g. `:lib-1:sub-lib` are not yet supported.
   * `API_FILE` - a file describing the public API for each artifact. These are generated using the [Metalava plugin](https://github.com/tylerbwong/metalava-gradle/)
      * Each module passed in `MODULES` should have it's own API file
      * The workflow looks for the file inside each module with the structure: `<SOURCE_DIR>/<MODULE>/<API_FILE>`
   * `VERSION_FILE` - the file containing the version number for the libraries. 
      * This is expected to be a single file across all artifacts/modules in a repository.
      * The workflow looks for the file inside each module with the structure: `<SOURCE_DIR>/<VERSION_FILE>`
4. Updated the Github release step to commit updated API files for all modules


## How to test

I've updated the [`release.yml`](https://github.com/guardian/source-apps/pull/211/files?w=1#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34) for Source in this [companion PR](https://github.com/guardian/source-apps/pull/211).

1. Run [the "release" action](https://github.com/guardian/source-apps/actions/workflows/release.yml) on the Source repo after selecting the _"ab/release/versioning"_ branch.
2. Confirm that the workflow completes successfully (it can take _a while_, so go cook dinner while it's running)
3. Once it is complete, confirm that the PR linked above has a comment from the workflow with details
4. (Android dev) Copy the version number in the comment
5. (Android dev) Update the source version in Android news app's version catalog to this version
6. (Android dev) Confirm the Android news app builds successfully.

> [!note]
> The Github release changes for PRs from `main` are untested as yet. I'll test them soon after I merge the Source PR linked above.


[See here](https://github.com/guardian/source-apps/actions/runs/15906758893) for a sample run.
See here for a sample comment: https://github.com/guardian/source-apps/pull/211#issuecomment-3008968004

## How can we measure success?

<!-- Do you expect errors to decrease? Do you expect user journeys to be simplified? What can be used to prove this? A filtered view of logs or analytics, etc? -->

## Have we considered potential risks?

<!-- What are the potential risks and how can they be mitigated? Does an error require an alarm? Should user help, infosec, or legal be informed of this change? Is private information guarded? Do we need to add anything in the backlog? -->


## Pre merge checklist

- [ ] Remove branch ref from checkout steps for shell scripts
- [ ] Remove branch ref from checkout steps for reusable actions (init, sonatype release)

adi-gnm committed:
Comment out post version steps

Also rename scala references to gradle in release workflow

adi-gnm committed:
Bring in the full reusable workflow from source

https://github.com/guardian/source-apps/pull/211

adi-gnm committed:
Fix support for multi module, multi artifact repos

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -91,64 +110,183 @@ jobs:
   generate-version-update-commits:
-    name: 🎊 Test & Version
-    needs: init
+    name: 🤸🏽 Generate version update commits
```
Deliberately not including `test` here. Gradle/Android libraries have a variety of tests - jvm unit, instrumented android, screenshots, and end to end. This being a reusable workflow, the test responsibility is left to individual repos to set up, as needed, as a previous job before invoking this workflow.

rtyley replied:
Excluding the tests makes sense, see:

* https://github.com/guardian/gha-scala-library-release-workflow/issues/42

Could we keep emoji language consistent tho? `🎊` not `🤸🏽`! 🤣 

I suggest:

```suggestion
    name: 🎊 Versioning
```

Long job names get truncated in the GitHub Actions UI, so it's nice to go shorter. 'Versioning' is also a name that we could probably adopt on `gha-scala-library-release-workflow` too.

adi-gnm committed:
Extract version calculation into external script

Disabled following jobs to iterate quickly

adi-gnm committed:
Checkout the calculate version script separately.

adi-gnm committed:
Specify scripts directory instead of pattern for sparse checkout

adi-gnm committed:
Re-enable subsequent jobs

adi-gnm committed:
Move unsigned commit script to `create-unsigned-commit.sh`

adi-gnm committed:
Disable following steps

adi-gnm committed:
Update CODEOWNERS to Roberto + Android admins

adi-gnm committed:
Re-enable following steps

Also add more step names

adi-gnm commented on .github/scripts/create-unsigned-commit.sh:
```diff
@@ -0,0 +1,104 @@
+# in a json object for later use
+IFS=',' read -ra MOD_ARRAY <<< "$MODULES"
+
+API_FILES_JSON="["
```
The number of modules is dynamic so we can't use static defined outputs for each. Instead we create a json output and add each module's API file change details to it.

This is parsed using `jq` in the Github release step when committing the changes to `main`.

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -91,64 +114,97 @@ jobs:
+              exit 1
+            fi
+            cd "$SOURCE_DIR"
+          fi
```
This is reused in several places - it is used to switch to project directory in case of a monorepo.

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -91,64 +114,97 @@ jobs:
+          repository: guardian/gha-gradle-library-release-workflow
+          path: gha-gradle-library-release-workflow
+          ref: 'ab/update-to-support-gradle-and-monorepo-configs'
+          sparse-checkout: '.github/scripts'
```
This is to allow us to use scripts in `.sh` files in this reusable workflow repository. It helps keep the yml file cleaner by moving large scripts to their own files.

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -168,131 +224,129 @@ jobs:
       version_file_path: ${{ steps.create-commit.outputs.version_file_path }}
       version_file_release_sha: ${{ steps.create-commit.outputs.version_file_release_sha }}
       version_file_post_release_content: ${{ steps.create-commit.outputs.version_file_post_release_content }}
+      api_files: ${{ steps.create-commit.outputs.api_files }}
```
This output contains json formatted data for API file commits for all modules.

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -168,131 +224,129 @@ jobs:
+
+            for mod in "${MOD_ARRAY[@]}"; do
+              MOD=$(echo "${mod}" | xargs)
+              ./gradlew ":${MOD}:publishReleasePublicationToCustomRepository" -Prepo.local=$LOCAL_ARTIFACTS_STAGING_PATH
```
Iterate over modules and create artifacts. Publish the artifacts at `LOCAL_ARTIFACTS_STAGING_PATH` for use in later steps.

adi-gnm committed:
Reset fetch depth to 1 for signing

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -413,24 +479,30 @@ jobs:
   github-release:
     name: 🔒 Update GitHub
-    needs:  [init, push-release-commit, sign]
+    needs:  [init, push-release-commit, sign, sonatype-release]
```
Ensure that `github-release` runs after `sonatype-release` has finished. This ensures that we don't do a github release/comment if sonatype release fails.

It also delays commenting on PRs till after the PREVIEW release is actually published.

rtyley replied:
Yeah, that's fair enough - I made those two tasks parallel for slightly dubious reasons: mostly because it made them fit on the screen better and meant that more of the job name text was visible to see!

As for using parallelisation elsewhere, `gha-scala-library-release-workflow` has one possible improvement I haven't got round to pursuing yet:

* https://github.com/guardian/gha-scala-library-release-workflow/issues/18

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -91,64 +114,97 @@ jobs:
+        with:
+          repository: guardian/gha-gradle-library-release-workflow
+          path: gha-gradle-library-release-workflow
+          ref: 'ab/update-to-support-gradle-and-monorepo-configs'
```
@ab-gnm TODO: Delete this before merge

rtyley replied:
Unfortunately this is the fiddly part of factoring the bash scripts out of the `reusable-release.yml` file: we end up having to execute a `actions/checkout` for our own repository to get those bash files.

The worst part is that there isn't any simple way to specify that we want to use the same commit that this `reusable-release.yml` came from  - [`github.workflow_sha`](https://docs.github.com/en/actions/reference/accessing-contextual-information-about-workflow-runs#github-context) gives us the commit of the initiating caller workflow, not the reusable _callee_ one - so when we start work on a PR we have to specify `ref` with the branch name of our PR, and then remember to remove the `ref` setting before we merge the PR! 

```yaml
ref: 'ab/update-to-support-gradle-and-monorepo-configs'
```

Gah!

Some workarounds for this (including an [OIDC-wordaround](https://github.com/orgs/community/discussions/31054#discussioncomment-3495110), that heroically parses `job_workflow_repo_ref` out of the OIDC JWT) are discussed here:

* https://github.com/orgs/community/discussions/31054#discussioncomment-4373060

...but none of them are great and manually editing the workflow before merging is probably equally good.

Note that for [composite actions](https://docs.github.com/en/actions/tutorials/creating-a-composite-action) (rather than reusable workflows), we can [access `github.action_path`](https://github.com/orgs/community/discussions/32593#discussioncomment-3616375) from the [`github` context](https://docs.github.com/en/actions/reference/accessing-contextual-information-about-workflow-runs#github-context), and get files from our repo that way.

I've started a PR to break `gha-scala-library-release-workflow` into reusable-actions - hopefully several of these can be reused by your workflow! :

* https://github.com/guardian/gha-scala-library-release-workflow/pull/69

adi-gnm replied:
> I've started a PR to break `gha-scala-library-release-workflow` into reusable-actions - hopefully several of these can be reused by your workflow! :
> 
> * [Break workflow up into smaller reusable files gha-scala-library-release-workflow#69](https://github.com/guardian/gha-scala-library-release-workflow/pull/69)

I've updated this PR to use the smaller reusable files from your PR for `init` and `sonatype release`. I'll be able to reuse the `sign` job as well but I'm waiting for you to integrate that in your PR before I do.

The Github release job is a bit different - my workflow also pushes the API signature files. So that would not be directly reusable.
Similarly, the "push release commit" job is also different between the two workflows, so that's the other secure job that's not directly reusable.

Run with smaller reusable files: https://github.com/guardian/source-apps/actions/runs/15977919078

adi-gnm commented on .github/workflows/reusable-release.yml:
```diff
@@ -168,131 +224,129 @@ jobs:
+        with:
+          repository: guardian/gha-gradle-library-release-workflow
+          path: gha-gradle-library-release-workflow
+          ref: 'ab/update-to-support-gradle-and-monorepo-configs'
```
@ab-gnm TODO: Delete this before merge

rtyley reviewed (approved):
I don't work Mondays, and I haven't really finished reviewing this PR (which for me here is more a case of adding commentary), but I wanted to give my approval anyway so as not to be a blocker!

Note this draft `gha-scala-library-release-workflow` PR may be useful in supplying several reusable actions we can share:

* https://github.com/guardian/gha-scala-library-release-workflow/pull/69

adi-gnm committed:
Update job name to 🎊 Versioning

Co-authored-by: Roberto Tyley <52038+rtyley@users.noreply.github.com>

adi-gnm committed:
Replace `init` with reusable step from Roberto's PR on scala workflow

adi-gnm committed:
Replace `sonatype release` with reusable step from Roberto's PR on scala workflow

adi-gnm committed:
Update step name for outputs from `init`

adi-gnm committed:
Replace ⚙ with 🎊 in job name 🥺

adi-gnm committed:
Bring in changes from scala workflow, and adopt signing action

adi-gnm committed:
Extract versioning job to a reusable action

adi-gnm committed:
Fix build issues with correct step names

adi-gnm committed:
Move calculate-version.sh back inside action.yml

adi-gnm committed:
Use full path for action file in versioning step

adi-gnm committed:
Explicitly specify output in `versioning/action.yml`

adi-gnm committed:
Use dash separeted java version variable

adi-gnm committed:
Use dash separeted java version variable in versioning action

adi-gnm committed:
Upload unsigned artifacts t to artifacts instead of cache

adi-gnm committed:
Replace _ with - in outputs from `create-unsigned-commit.sh`

adi-gnm committed:
Replace _ with - in `release-type` input reads in `update-github` job

adi-gnm committed:
Comment out the committing of api and version files for main branch in update-github. These are already committed in the sign job.

Also remove commented sections from unused reusable actions in scala workflow.

adi-gnm committed:
Move "push-release-commit" into its own action.

adi-gnm committed:
Move "create-artifacts" into its own action.

adi-gnm committed:
Move "update-github" into its own action.

adi-gnm committed:
Add logging to identify failure step

adi-gnm committed:
More logging

adi-gnm committed:
More logging

adi-gnm committed:
Don't go back to base path before committing version file

adi-gnm committed:
Remove commented base path changes

adi-gnm committed:
Remove logging

adi-gnm committed:
Use release commit id instead of version tag to checkout.

Version tag isn't pushed till later.

adi-gnm committed:
Provide java version

adi-gnm committed:
Specify shell for update-github steps

adi-gnm committed:
Adding logging

adi-gnm committed:
Expand full reference in log

adi-gnm committed:
Extract reference to a variable

adi-gnm committed:
Set token and repo env variables before deleting temporary branch

adi-gnm committed:
Remove logging

adi-gnm committed:
Remove `LOCAL_ARTIFACTS_STAGING_PATH` from environment as it's only used inside `create-artifacts`.

We do need a way of ensuring that the path is consistent with the one used in `sign` and `sonatype-release` actions in the scala workflow.

adi-gnm committed:
Remove `TEMPORARY_BRANCH` environment variable.

We now create it inside `push-release-commit`, and provide it as an output to be used in `update-github`

adi-gnm committed:
Use artifacts instead of cache for transferring unsigned repo between `versioning` and `push-release-commit`

adi-gnm committed:
Don't switch into temp repo directory

adi-gnm committed:
Replace PR branch name with main
